### DISCUSSION TITLE
Exercise: Solve 2 small problems in real estate using blockchain (Smart Tenancy Contract)

### DISCUSSION THREAD

**Discussion by CHOMNANP** [web3together/web3together#10] [Ideas]
The objective is to use the problems as an exercise to practice using blockchain.

Title: Smart Tenancy Contract
Description: 
"Real estate agency ABC" (R.ABC) was exclusively authorized to sell and rent all units in Build A (B.A).
This means that the R.ABC will earn certain amount of fee for any sale or rental transaction in the B.A.

R.ABC does have 1 sale manager, 2 sale supervisors, 4 sale staffs. R.ABC also give sub contract to 5 other sale agencies (Sub.A5) which has total of 120 sale persons. R.ABC sale person is the contact person for Sub.A5. Any sale or rent from Sub.A5 will go through R.ABC sale staffs.

Here are the fee structure

### Problem 1 (Fee Scheme)
**Sale transaction scheme**
````
Unit Owner: 90% of total sale amount
Building Owner Fee: 2% of total sale amount
Agency Fee: 8% of total sale amount
````

**Rent transaction scheme (This is monthly fee scheme)**
````
Unit Owner: 85% of total rent amount
Building Owner Fee: 5% of total rent amount
Agency Fee: 10% of total rent amount
````

**Fee Scheme 1: Sale or rent by R.ABC sale team**
````
R.ABC Company Fee: 40% of total agency fee
R.ABC sale team: 10% of total agency fee
R.ABC sale manager: 5% of total agency fee
R.ABC sale supervisor: 5% of total agency fee
R.ABC sale staff: 40% of total agency fee
````

**Fee Scheme 2: Sale or rent by Sub.A5**
````
R.ABC Company Fee: 40% of total agency fee
R.ABC sale team: 10% of total agency fee
R.ABC sale manager: 5% of total agency fee
R.ABC sale supervisor: 5% of total agency fee
R.ABC sale staff: 10% of total agency fee
Sub.A5: 30% of total agency fee
````

### Problem 2 (Payments)

The agency fee will be hold until the payment has meet certain step in the rule. A sale of a unit will follow one the rule bellow. 
Note: This rules are only applied to sale agency fee. All rent agency fee will be released once payment is complete 100%.

````
Sale Rule #1:
#1. Customer make 10% of sale payment => Release of 15% of agency fee
#2. Customer make 30% of sale payment => Release of 20% of agency fee
#3. Customer make 50% of sale payment => Release of 20% of agency fee
#4. Customer make 70% of sale payment => Release of 20% of agency fee
#5. Customer make 100% of sale payment => Release of 25% of agency fee
````

````
Sale Rule #2:
There are 24 installments for the sale of the unit in B.A.
#1. Customer make payment #1 => Release of 20% of agency fee
#2. Customer make payment #10 => Release of 20% of agency fee
#3. Customer make payment #15 => Release of 20% of agency fee
#4. Customer make payment #20=> Release of 20% of agency fee
#5. Customer make payment #24 => Release of 20% of agency fee
````

Based on these problems, let's apply smart contract to solve this specific case.

**Comment by CHOMNANP** [comment]
Can any propose few blockchain technology to solve these problems?
I would like to suggest using the private network blockchain.

**Reply by BunHouth** [reply]
### Deploy Network
In that case we can using [__ganache__](https://github.com/trufflesuite/ganache) for local development. 

Using testnet for public chain.
[Goerli](https://goerli.etherscan.io/)
Get Testnet ETH [here](https://faucet.goerli.mudit.blog/)

Using remote RPC
[infura](https://infura.io/)
[alchemy](http://alchemy.com/)

### Solidity on local development library

1.  [Truffle](https://www.trufflesuite.com/docs/truffle/getting-started/installation)
2. [Hardhat](https://hardhat.org/getting-started/)

### Solidity local development on cloud 
1. [Remix](https://remix.ethereum.org/)

**Reply by samnang** [reply]
@BunHouth thanks for putting a good list.

**Comment by samnang** [comment]
@CHOMNANP Could you help to clarify the requirements:

1. Who are the actors in the application?
2. How does the agency get authorized from the owner of their assets?
3. How does the agency manage their team?
4. Who will be the one that initiates the transactions?

For local blockchain node in development, you could try [Ganache](https://www.trufflesuite.com/ganache) or [Hardhat Network](https://hardhat.org/hardhat-network/).

**Reply by CHOMNANP** [reply]
Hi bong @samnang, I was thinking of using blockchain as a sub system of another system.

The concept is allow users to interact to App/Web Application via an RESS(API), so that we can manage the user experiences. Then the RESS will manage all security to access and perform transaction in blockchain on behalf of the user.
User does not have direct access to blockchain.

![image](https://user-images.githubusercontent.com/5925050/143835283-ad1dde07-9edf-4295-909a-60ab63d96e89.png)

````
Q1: Who are the actors in the application?
A1: Sale staff will create sale/rent transaction and choose payment term & rule as customer agree on the signed sale/rent contract. Cashier staff will create payment for each sale/rent. 
````
````
Q2: How does the agency get authorized from the owner of their assets?
A2: The authorization is done through the custom process in the RESS(refer to the image above). This is mean RESS will have UI for admin to manage the agency.
````
````
Q3: How does the agency manage their team?
A3: Agency will provide list of their team to the system admin. System admin will create and update the team information accordingly. This is done in RESS.
````
````
Q4: Who will be the one that initiates the transactions?
A4: Sale staff will create the sale/rent transaction. Cashier staff will create the payment for each sale/rent transaction. This is done through the RESS (refer to the image above).
````

**Reply by BunHouth** [reply]
@CHOMNANP I think to achieve that we need to develop a smart contract which include those function first, then we can develop rest api and app/web.

**Reply by CHOMNANP** [reply]
@BunHouth, totally agree. 

We don't need to develop a full function app.

We can leave "App/Web Application" and "RESS) apart first. We can do the blockhain part first. We can use cmd or any blockchain client tool to test out the blockchain part.

**Reply by samnang** [reply]
If it's centralized and is only accessed by RSS, we need to pay attention why do we even need blockchain? Are you any other parties involved other than RSS?

![image](https://user-images.githubusercontent.com/131172/143992212-80c62065-dd7e-4241-90c1-f07742270257.png)

**Reply by BunHouth** [reply]
using truffle or hardhat is JS friendly and easy to test.

**Reply by BunHouth** [reply]
I think working in blockchain we don't need to use with Rest.  Only authorize wallet that able to perform any action so security should be solve.

**Reply by CHOMNANP** [reply]
Here's my opinion
- Earn trust from investor and customer because we have a proof that all sales, rents and payment transaction are store on a very secure platform. All information that is written regarding sale, rent and payment cannot be changed.
- Could be just good case to learn blockchain and practice smart contract

**Reply by BunHouth** [reply]
this repo also good for testing [FlightSurety](https://github.com/jensendarren/FlightSurety)

**Reply by csokun** [reply]
If you want to learn Smart Contract try something simple scope.

As a landlord owner, I should be able to put my property on the market (rent).

As a tenant, I should be able to submit an expression of interest against a listed property by locking my token for a period of time.

Binding Contract:
Once a tenant submitted an expression of interest,
 - reject if the sender does hold sufficient balance
 - on successful submission, the sender balance will be locked for 24 hrs.
